Dalamud/Game/Gui/ContextMenu/ContextMenuGui.cs

```diff
@@ -0,0 +1,501 @@
+        this.raptureAtkModuleOpenAddon = Marshal.GetDelegateForFunctionPointer<RaptureAtkModuleOpenAddonDelegate>(this.address.RaptureAtkModuleOpenAddon);
+
+        this.contextAddonOpenByAgentHook.Enable();
+        this.contextMenuOnVf72Hook.Enable();
+    }
+
+    private unsafe delegate nint ContextAddonOpenByAgentDelegate(RaptureAtkModule* module, byte* addonName, AtkUnitBase* addon, int valueCount, AtkValue* values, AgentInterface* agent, nint a7, ushort parentAddonId);
+    
+    private unsafe delegate bool ContextMenuOnVf72Delegate(AddonContextMenu* addon, int selectedIdx, byte a3);
+    
+    private unsafe delegate nint RaptureAtkModuleOpenAddonDelegate(RaptureAtkModule* a1, uint addonNameId, uint valueCount, AtkValue* values, AgentInterface* parentAgent, long unk, uint ownerAddonId, int unk2);
+
+    /// <inheritdoc/>
+    public event IContextMenu.OnMenuOpenedDelegate OnMenuOpened;
+
+    private Dictionary<ContextMenuType, List<MenuItem>> MenuItems { get; } = new();
+
+    private AgentInterface* SelectedAgent { get; set; }
+
+    private ContextMenuType? SelectedMenuType { get; set; }
+
+    private List<MenuItem>? SelectedItems { get; set; }
+
+    private HashSet<nint> SelectedEventInterfaces { get; } = new();
+
+    private AtkUnitBase* SelectedParentAddon { get; set; }
+
+    // -1 -> -inf: native items
+    // 0 -> inf: selected items
+    private List<int> MenuCallbackIds { get; } = new();
+
+    private IReadOnlyList<MenuItem>? SubmenuItems { get; set; }
+
+    /// <inheritdoc/>
+    public void Dispose()
+    {
+        var manager = RaptureAtkUnitManager.Instance();
+        var menu = manager->GetAddonByName("ContextMenu");
+        var submenu = manager->GetAddonByName("AddonContextSub");
+        if (menu->IsVisible)
+            menu->FireCallbackInt(-1);
+        if (submenu->IsVisible)
+            submenu->FireCallbackInt(-1);
+
+        this.contextAddonOpenByAgentHook.Dispose();
+        this.contextMenuOnVf72Hook.Dispose();
+    }
+
+    /// <inheritdoc/>
+    public void AddMenuItem(ContextMenuType menuType, MenuItem item)
+    {
+        if (!this.MenuItems.TryGetValue(menuType, out var items))
+            this.MenuItems[menuType] = items = new();
+        items.Add(item);
+    }
+
+    /// <inheritdoc/>
+    public bool RemoveMenuItem(ContextMenuType menuType, MenuItem item)
+    {
+        if (!this.MenuItems.TryGetValue(menuType, out var items))
+            return false;
+        return items.Remove(item);
+    }
+
+    private AtkValue* ExpandContextMenuArray(Span<AtkValue> oldValues, int newSize)
+    {
+        // if the array has enough room, don't reallocate
+        if (oldValues.Length >= newSize)
+            return (AtkValue*)Unsafe.AsPointer(ref oldValues[0]);
+
+        var size = (sizeof(AtkValue) * newSize) + 8;
+        var newArray = (nint)IMemorySpace.GetUISpace()->Malloc((ulong)size, 0);
+        NativeMemory.Fill((void*)newArray, (nuint)size, 0);
```

Comment:
Probably a good idea to throw a `OutOfMemoryException` first if `newArray` is null, just in case future modification to this class results in `newSize` value that is not sane.